### PR TITLE
sql: using catalog reader, allow descriptor scanning by span

### LINKED ISSUE #130997: List multiple table names in the Hot Ranges table for coalesced ranges containing different tables
**Is your feature request related to a problem? Please describe.**

A recent investigation into an issue revealed that It's difficult to trace the root cause of a hot range when a range is a coalesced range comprising of many different tables.

**Describe the solution you'd like**

Need better observability to show the names of various tables that are part of the same range in the Hot Ranges table when those ranges have been coalesced and include data from different tables. This will be useful for identifying which tables are contributing to the hot ranges when ranges span multiple tables.

**Describe alternatives you've considered**

On a remote session with the customer, we used LogSpy to look at the requests that were hitting the range from a single IP address. The IP was associated with an haproxy instance, and so the traffic was coming in from an external source.

After analyzing the sessions and noting the IP and Port from LogSpy output, we found that it was running only SQL to a different table. Upon checking the range's  keyspan, we saw that the range was a coalesced range comprising of many different tables, including the table that was getting the brunt of the load. 

Jira issue: CRDB-42328

Epic CRDB-43151

### PR DESCRIPTION
**Pull request by angles-n-daemons**
Summary: new `catalogReader.GetDescriptorsInSpans(..., spans []roachpb.Span)` function which acts as named.

-----------
sql: using catalog reader, allow descriptor scanning by span

Historically, to estimate which tables and indexes existed within the
span of a range, we used the Collection's `GetAllDescriptors` function
and returned the values which fell in between the start and end keys of
the range's span. This approach had the benefit of being precise, but
the drawback of being computationally expensive - since the system can
have hundreds of nodes and theoretically millions of tables, using
`GetAllDescriptors` begins to become a bottleneck.

This approach was modified so that instead of pulling all descriptors
for the system when computing range contents, the system instead used
the start key of the range to identify the exact table + index at that
point within the keyspace
([change](https://github.com/cockroachdb/cockroach/pull/77277/files)).
This works if each table, index has their own range, but quickly breaks
down if tables share a range. Consider the following layout of data:

```
T1=Table 1
T2=Table 2
R1=Range1

└─────T1─────┴─────T2────┴─────T3─────┘
└────────┴─────────R1───────┴─────────┘
```

Since the start key of the range falls with Table 1, the system
associates the range with only Table 1, despite it containing Tables 2
and 3.

Using this information, it becomes necessary to identify a set of
descriptors within a certain span. This PR introduces the
`ScanDescriptorsInSpans` function which does just that, allows the user
to specify a set of spans whose descriptors are important and then
return a catalog including those descriptors.

It does this by translating the span keys into descriptor span keys and
scanning them from the descriptors table. For example given a span
`[/Table/Users/PKEY/1, /Table/Users/SECONDARY/chicago]` where the ID for
the Users table is `5`, it will generate a descriptor span
`[/Table/Descriptors/PKEY/5, /Table/Descriptors/PKEY/6]`.

This descriptor too comes with its drawbacks in that within the
descriptor space, keys are scoped by table, and not necessarily indexes.
That means in a following PR, the status server will be responsible for
taking these descriptors, which include all indexes in the tables
pulled, and filtering it down to only the indexes which appear in the
specified range.

The bulk of the changeset is in updating the datadriven tests to test
this behavior, the primary area of focus for review should be the
`pkg/sql/catalog/internal/catkv/catalog_reader.go` file (~75 LOC).

Epic: CRDB-43151
Fixes: #130997

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/internal/catkv/catalog_reader_cached.go [hunk 1]
```diff
@@ -220,6 +221,13 @@ func (c *cachedCatalogReader) ScanAll(ctx context.Context, txn *kv.Txn) (nstree.
 	return read, nil
 }
 
+// ScanDescriptorsInSpans is part of the CatalogReader interface.
+func (c *cachedCatalogReader) ScanDescriptorsInSpans(
```
Comment (angles-n-daemons) [inline pkg/sql/catalog/internal/catkv/catalog_reader_cached.go:225]: Unsure of a good caching strategy and when cache invalidation occurs, looking for input if any

**Comment by angles-n-daemons** [comment]
I expect this approach to scale well to ~10⁴ ranges for a single Node. For HotRanges specifically, it may be worthwhile scanning in batches of a bit less than that size to accomplish the task.

**Comment by angles-n-daemons** [comment]
This approach may also increase latency on the existing hot ranges page. Might require some testing.

File: pkg/sql/catalog/internal/catkv/catalog_reader.go [hunk 2]
```diff
@@ -158,6 +161,76 @@ func (cr catalogReader) ScanAll(ctx context.Context, txn *kv.Txn) (nstree.Catalo
...
+		if err != nil {
+			return mc.Catalog, err
+		}
+		descSpans[i] = descSpan
+	}
+
+	cq := catalogQuery{codec: cr.codec}
+	err := cq.query(ctx, txn, &mc, func(codec keys.SQLCodec, b *kv.Batch) {
+		for _, descSpan := range descSpans {
+			scanRange(ctx, b, descSpan.Key, descSpan.EndKey)
```
Comment (angles-n-daemons) [inline pkg/sql/catalog/internal/catkv/catalog_reader.go:224]: Is this de-duplicated?

Comment (angles-n-daemons) [reply pkg/sql/catalog/internal/catkv/catalog_reader.go:224]: better yet, add a test

**Review by dhartunian — COMMENTED** [review]
> Reviewed 1 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133190)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons)

___
*[`-- commits` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlXaP-zVOvXODC--D):*
nit: do you mean "descriptor" here?
___
*[`-- commits` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlak33aFaKAO1C1s1):*
also "descriptor" here?
___
*[`-- commits` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlmIr9icPAqDYU9l6):*
Should this be `/Table/5/PKEY/1`? We have the ID in the keys, not the name I believe.
___
*[`-- commits` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUmnc724bH1dP1ZOq1):*
I think labeling this work as benefiting the hot spots epic is reasonable.
___
*[`pkg/sql/catalog/internal/catkv/catalog_reader_cached.go` line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133190#gh-1811295888:-OAXqcjdAYIN-zyfYNHW:bmbjla1) ([raw file](https://github.com/cockroachdb/cockroach/blob/76cc517c3d7af165356accaed7a5b053b9ec2e0d/pkg/sql/catalog/internal/catkv/catalog_reader_cached.go#L225)):*
<details><summary><i>Previously, angles-n-daemons (Brian Dillmann) wrote…</i></summary><blockquote>

Unsure of a good caching strategy and when cache invalidation occurs, looking for input if any
</blockquote></details>

My naive guess: cached catalog reader holds a mutable catalog. We can maintain a record of which spans have been added to the catalog already and skip scanning them on subsequent requests.
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_app` line 727 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXt5jU8CR2ZLXfh4_0:-OAXt5jU8CR2ZLXfh4_1:bi34tvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_app#L727)):*
> ```
> 
> # start in the middle of the first table
> scan_descriptors_in_span start=110/1/1 end=112/1
> ```

this test repeats the prior one
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_system` line 622 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXw7y05vrhLIOVclXu:-OAXw7y11ygnPSsE3YLL:bx2z5xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_system#L622)):*
> ```
> # disabled because it panics if the same key is passed
> # scan_descriptors_in_span start=110/1 end=110/1
> # ----
> ```

ack. I see how this requires some modification to the test harness. consider filing a separate ticket to follow up for these special cases.
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_system` line 745 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXu4SD7vJNh3YIZ-Rg:-OAXu4SD7vJNh3YIZ-Rh:b-8f15ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_system#L745)):*
> ```
> 
> # start in the middle of the first table
> scan_descriptors_in_span start=110/1/1 end=112/1
> ```

same here, test repeats.

**Review by angles-n-daemons — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133190)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`-- commits` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlXaP-zVOvXODC--D):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: do you mean "descriptor" here?
</blockquote></details>

oy, yes I did - changed.
___
*[`-- commits` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlak33aFaKAO1C1s1):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

also "descriptor" here?
</blockquote></details>

done.
___
*[`-- commits` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlmIr9icPAqDYU9l6):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Should this be `/Table/5/PKEY/1`? We have the ID in the keys, not the name I believe.
</blockquote></details>

It could be, it was mostly illustrative. How do you think it reads? The string format of the key would be `/Table/5/1/1`.
___
*[`-- commits` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUmnc724bH1dP1ZOq1):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think labeling this work as benefiting the hot spots epic is reasonable.
</blockquote></details>

done.
___
*[`pkg/sql/catalog/internal/catkv/catalog_reader_cached.go` line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133190#gh-1811295888:-OAdWeMPASAvVGL1zrcO:b5b7i8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/76cc517c3d7af165356accaed7a5b053b9ec2e0d/pkg/sql/catalog/internal/catkv/catalog_reader_cached.go#L225)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

My naive guess: cached catalog reader holds a mutable catalog. We can maintain a record of which spans have been added to the catalog already and skip scanning them on subsequent requests.
</blockquote></details>

Can I add the caching behavior in a follow up pr?
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_app` line 727 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXt5jU8CR2ZLXfh4_0:-OAdXBVn9WL24O8GI40E:b-x0vep) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_app#L727)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

this test repeats the prior one
</blockquote></details>

nice catch, removed.
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_system` line 622 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXw7y05vrhLIOVclXu:-OAdXDek2P1rIg9jy2ai:buxseri) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_system#L622)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

ack. I see how this requires some modification to the test harness. consider filing a separate ticket to follow up for these special cases.
</blockquote></details>

https://cockroachlabs.atlassian.net/browse/CRDB-43984?atlOrigin=eyJpIjoiNWY0YjA3YjFjYjM3NDEwY2I5YTM3NGU3MWUxNGRhYmUiLCJwIjoiaiJ9
___
*[`pkg/sql/catalog/internal/catkv/testdata/testdata_system` line 745 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAXu4SD7vJNh3YIZ-Rg:-OAdX9Pt5yStCy2w8Eer:brpmqc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b9ee03fba37f314771d8624b2619f10b5bbb59/pkg/sql/catalog/internal/catkv/testdata/testdata_system#L745)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

same here, test repeats.
</blockquote></details>

removed.

**Review by dhartunian — APPROVED** [review]
:lgtm: just some nits

> Reviewed 1 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-:-OB26Xc4Dactk6voyUDM:b-yl2rbz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons)

___
*[`-- commits` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OAUlmIr9icPAqDYU9l6):*
<details><summary><i>Previously, angles-n-daemons (Brian Dillmann) wrote…</i></summary><blockquote>

It could be, it was mostly illustrative. How do you think it reads? The string format of the key would be `/Table/5/1/1`.
</blockquote></details>

Gotcha. fine to keep :thumbsup: thx
___
*[`-- commits` line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133190#-OB25yc8-uZjZQ0BBvGy):*
I'm pretty sure you want `#130997` :smile: 
___
*[`pkg/sql/catalog/internal/catkv/catalog_reader_cached.go` line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133190#gh-1811295888:-OB26FiT4BGjTE8ExhDj:b-sq4o33) ([raw file](https://github.com/cockroachdb/cockroach/blob/76cc517c3d7af165356accaed7a5b053b9ec2e0d/pkg/sql/catalog/internal/catkv/catalog_reader_cached.go#L225)):*
<details><summary><i>Previously, angles-n-daemons (Brian Dillmann) wrote…</i></summary><blockquote>

Can I add the caching behavior in a follow up pr?
</blockquote></details>

Yes. Can you add a docstring here that there's no caching implemented and a TODO comment that links to the issue.

**Pull request merged**